### DISCUSSION TITLE
How to get class instance here?

### DISCUSSION THREAD

**Discussion by EduModsS** [vfsfitvnm/frida-il2cpp-bridge#376] [Q&A]
// Assembly-CSharp
class MovePlayer : UnityEngine.MonoBehaviour
{ ... System.Void StartBeeMode(); // 0x0138c4b8 ... }

eu tentei obter com Il2Cpp.gc.choose, and without success, the only way was to hook a method for it to call (this), however, I would like these methods (method.implementation) to return the original value. or that it is possible get class instance to invoke / edit methods without hooking method, help me

**Comment by bluewave41** [comment]
There are 2 ways to scan for objects: https://github.com/vfsfitvnm/frida-il2cpp-bridge/wiki/Snippets#heap-scanning

If you haven't, make sure to try both of them.

> however, I would like these methods (method.implementation) to return the original value

So return the original value and save the reference.

```
assembly.class('MovePlayer').method('StartBeeMode').implementation = function() {
    savedReference = this;
    return this.method('StartBeeMode').invoke();
}

**Comment by EduModsS** [comment]
ok, the second one worked, the first one I didn't understand correctly

**Comment by bluewave41** [comment]
> ok, the second one worked, the first one I didn't understand correctly

To scan for objects you should be doing something like

```
    const assembly.= Il2Cpp.domain.assembly('Assembly-CSharp').image;
    const instances = Il2Cpp.gc.choose(assembly.class('MovePlayer'))
```

`instances` at this point will be an array of all the allocated `MovePlayer` classes

I've never had to use the heap method but going by the docs it should be something like

```
    const assembly = Il2Cpp.domain.assembly('Assembly-CSharp').image;
    const snapshot = Il2Cpp.MemorySnapshot.capture();
    const instances = snapshot.objects.filter(Il2Cpp.isExactly(assembly.class('MovePlayer')));
    snapshot.free();
```

**Reply by EduModsS** [reply]
I will test this soon

**Comment by EduModsS** [comment]
and also, is there a way to disable the implementation after that? example:

blablabla.implementation = null;

**Reply by vfsfitvnm** [reply]
`blablabla.revert()`

**Comment by EduModsS** [comment]
ok, everything went well, last question, for the blabla.revert i need to call the arguments of method yet?

**Reply by commonuserlol** [reply]
no

**Reply by vfsfitvnm** [reply]
no, `Il2Cpp.Method::revert` restores the original implementation, that is your custom one (if any) gets deleted

**Comment by EduModsS** [comment]
oh ok, thanks, everything worked correctly
